include/nbl/builtin/hlsl/impl/tgmath_impl.hlsl

```diff
@@ -0,0 +1,767 @@
+#ifndef _NBL_BUILTIN_HLSL_TGMATH_IMPL_INCLUDED_
+#define _NBL_BUILTIN_HLSL_TGMATH_IMPL_INCLUDED_
+
+#include <nbl/builtin/hlsl/cpp_compat/basic.h>
+#include <nbl/builtin/hlsl/concepts.hlsl>
+#include <nbl/builtin/hlsl/concepts/vector.hlsl>
+#include <nbl/builtin/hlsl/spirv_intrinsics/core.hlsl>
+#include <nbl/builtin/hlsl/spirv_intrinsics/glsl.std.450.hlsl>
+#include <nbl/builtin/hlsl/ieee754.hlsl>
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace tgmath_impl
+{
+
+template<typename T, typename U NBL_STRUCT_CONSTRAINABLE>
+struct lerp_helper;
+
+#ifdef __HLSL_VERSION
+#define MIX_FUNCTION spirv::fMix
+#else
+#define MIX_FUNCTION glm::mix
+#endif
+
+#define DEFINE_LERP_HELPER_COMMON_SPECIALIZATION(TYPE)\
+template<>\
+struct lerp_helper<TYPE, TYPE>\
+{\
+	static inline TYPE __call(NBL_CONST_REF_ARG(TYPE) x, NBL_CONST_REF_ARG(TYPE) y, NBL_CONST_REF_ARG(TYPE) a)\
+	{\
+		return MIX_FUNCTION(x, y, a);\
+	}\
+};\
+\
+template<int N>\
+struct lerp_helper<vector<TYPE, N>, vector<TYPE, N> >\
+{\
+	static inline vector<TYPE, N> __call(NBL_CONST_REF_ARG(vector<TYPE, N>) x, NBL_CONST_REF_ARG(vector<TYPE, N>) y, NBL_CONST_REF_ARG(vector<TYPE, N>) a)\
+	{\
+		return MIX_FUNCTION(x, y, a);\
+	}\
+};\
+\
+template<int N>\
+struct lerp_helper<vector<TYPE, N>, TYPE>\
+{\
+	static inline vector<TYPE, N> __call(NBL_CONST_REF_ARG(vector<TYPE, N>) x, NBL_CONST_REF_ARG(vector<TYPE, N>) y, NBL_CONST_REF_ARG(TYPE) a)\
+	{\
+		return MIX_FUNCTION(x, y, a);\
+	}\
+};\
+
+DEFINE_LERP_HELPER_COMMON_SPECIALIZATION(float32_t)
+DEFINE_LERP_HELPER_COMMON_SPECIALIZATION(float64_t)
+
+#undef DEFINE_LERP_HELPER_COMMON_SPECIALIZATION
+#undef MIX_FUNCTION
+
+// LERP
+
+template<typename T>
+struct lerp_helper<T, bool>
+{
+	static inline T __call(NBL_CONST_REF_ARG(T) x, NBL_CONST_REF_ARG(T) y, NBL_CONST_REF_ARG(bool) a)
+	{
+		if (a)
+			return y;
+		else
+			return x;
+	}
+};
+
+template<typename T, int N>
+struct lerp_helper<vector<T, N>, vector<bool, N> >
+{
+	using output_vec_t = vector<T, N>;
+
+	static inline output_vec_t __call(NBL_CONST_REF_ARG(output_vec_t) x, NBL_CONST_REF_ARG(output_vec_t) y, NBL_CONST_REF_ARG(vector<bool, N>) a)
+	{
+		output_vec_t retval;
+		for (uint32_t i = 0; i < vector_traits<output_vec_t>::Dimension; i++)
+			retval[i] = a[i] ? y[i] : x[i];
+		return retval;
+	}
+};
+
+// ISNAN
+
+template<typename UnsignedInteger NBL_FUNC_REQUIRES(hlsl::is_integral_v<UnsignedInteger> && hlsl::is_unsigned_v<UnsignedInteger>)
+inline bool isnan_uint_impl(UnsignedInteger val)
+{
+	using AsFloat = typename float_of_size<sizeof(UnsignedInteger)>::type;
+	return bool((ieee754::extractBiasedExponent<UnsignedInteger>(val) == ieee754::traits<AsFloat>::specialValueExp) && (val & ieee754::traits<AsFloat>::mantissaMask));
```

Comment:
there's a faster check:
1. AND with `(numeric_limits<UnsignedInteger>::max>>1)` - clear the sign mask
2. check for `> (specialValueExp<<mantissaBits)`